Add more inplace arguments to quantization top level API

dzhulgakov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26782 Add more inplace arguments to quantization top level API**

At least we should be consistent on top-level APIs and prepare/convert/etc.

Logic is inplace=False by default but top-level APIs take care of doing fewer copies.

Also renames always-inplace methods like add_observer to have underscore in the end.

One fix for MinMaxObserver was triggered by deepcopy surfacing that we were accidentally keeping autograd around

Differential Revision: [D17595956](https://our.internmc.facebook.com/intern/diff/D17595956)

Commit:
Add more inplace arguments to quantization top level API

[ghstack-poisoned]

Commit:
Update on "Add more inplace arguments to quantization top level API"


At least we should be consistent on top-level APIs. I'm less sure about prepare/convert/etc. If we add inplace=True to them it's kind of useless (user can call deep copy). If we add inplace=False - it's easy to end up with too many copies and it doesn't fit regular workflow of the user.

Maybe we should add argument only to `prepare` - I'm open to suggestions

[ghstack-poisoned]

ZolotukhinM commented on torch/quantization/quantize.py:
```diff
@@ -176,20 +177,33 @@ def add_quant_dequant(module):
+            configuration, qconfig applies to all submodules of a given
+            module unless qconfig for the submodules are specified (when the
+            submodule already has qconfig attribute)
+        inplace: carry out model transformations in-place, the original model is destroyed
```
Nit: is "destroyed" the right word here? It might look to users that if I do `prepare(model)` then the `model` is destroyed and I cannot use it anymore.

ZolotukhinM reviewed (commented):
Looks good to me, but I'll let other people to review it.

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -241,19 +255,22 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING):
+        model = copy.deepcopy(model)
     model.eval()
-    prepare(model)
+    model = prepare(model, inplace=True)
```
why do we need to assign with `inplace=True`?

dzhulgakov replied:
just for consistency with potential graph mode

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -176,20 +177,33 @@ def add_quant_dequant(module):
+        model = copy.deepcopy(model)
+    propagate_qconfig_(model)
+    # sanity check common API misusage
+    if not any(hasattr(m, 'qconfig') and m.qconfig for m in model.modules()):
```
although this is a warning, but maybe we want to check all the modules in the module hierarchy to say there is no qconfig applied?

dzhulgakov replied:
that's what it does. .modules() returns all submodules recursively

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -290,17 +308,19 @@ def quantize_dynamic(model, qconfig_dict=None, dtype=torch.qint8, mapping=DEFAUL
-    propagate_qconfig(model, qconfig_dict)
-    convert(model, mapping)
+    propagate_qconfig_(model, qconfig_dict)
+    model = convert(model, mapping, inplace=True)
```
why do we need to return with inplace=True?

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -59,25 +59,33 @@ def checkQuantized(model):
+
+        # in-place version
+        model = SingleLayerLinearModel()
+        quantize(model, test_only_eval_fn, self.calib_data, inplace=True)
```
for`inplace=True` should we either always assign or not assign the result?

dzhulgakov replied:
good catch, I might have missed a few places

dzhulgakov commented:
Actually, I think I'll keep unittests with out assignment if inplace=True. It should be doable with graph mode too (by assigning the graph in-place) and APIs look much nicer without the need for extra assignemnt

Commit:
Update on "Add more inplace arguments to quantization top level API"


At least we should be consistent on top-level APIs and prepare/convert/etc.

Logic is inplace=False by default but top-level APIs take care of doing fewer copies.

Also renames always-inplace methods like add_observer to have underscore in the end.

One fix for MinMaxObserver was triggered by deepcopy surfacing that we were accidentally keeping autograd around

[ghstack-poisoned]

raghuramank100 commented:
> Stack from [ghstack](https://github.com/ezyang/ghstack):
> 
> * **#26782 Add more inplace arguments to quantization top level API**
> 
> At least we should be consistent on top-level APIs and prepare/convert/etc.
> 
> Logic is inplace=False by default but top-level APIs take care of doing fewer copies.
> 
> Also renames always-inplace methods like add_observer to have underscore in the end.
> 
> One fix for MinMaxObserver was triggered by deepcopy surfacing that we were accidentally keeping autograd around

The minmaxobserver fix also showed up as part of QAT debug, but my diffs havent landed yet, all observers now have with torch.no_grad() or detach to prevent this.

jerryzh168 commented on test/test_quantized_nn_mods.py:
```diff
@@ -300,11 +300,11 @@ def test_linear_api(self, batch_size, in_features, out_features, use_bias, use_f
         # Sequential allows swapping using "convert".
         quantized_float_linear = torch.nn.Sequential(float_linear)
-        torch.quantization.convert(quantized_float_linear)
+        quantized_float_linear = torch.quantization.convert(quantized_float_linear, inplace=True)
```
one missed assignment here

masahi commented on torch/quantization/quantize.py:
```diff
@@ -176,20 +177,33 @@ def add_quant_dequant(module):
-    add_observer(model)
+    if not inplace:
+        model = copy.deepcopy(model)
+    propagate_qconfig_(model)
```
why qconfig_dict is not passed to propagate_qconfig_(...)? qconfig_dict is not used inside this function.
Is it intended?
